src/utils/PopupPositioning.js

```diff
@@ -0,0 +1,225 @@
+      toReturn.left = flowRight ? -bodyRect.width : -activatorRect.width;
+      break;
+    case FlyoutMenuDirection.LeftStart:
+      toReturn.top = flowDown ? activatorRect.height : -bodyRect.height;
+      toReturn.left = flowRight ? -bodyRect.width : -activatorRect.width;
+      break;
+    case FlyoutMenuDirection.RightEnd:
+      toReturn.top = flowDown ? bodyRect.height : -activatorRect.height;
+      toReturn.left = flowRight ? -activatorRect.width : -bodyRect.width;
+      break;
+    case FlyoutMenuDirection.RightStart:
+      toReturn.top = flowDown ? activatorRect.height : -bodyRect.height;
+      toReturn.left = flowRight ? -activatorRect.width : -bodyRect.width;
+      break;
+    case FlyoutMenuDirection.BottomEnd:
+      toReturn.top = flowDown ? 0 : -(activatorRect.height + bodyRect.height);
+      toReturn.left = flowRight ? bodyRect.width - activatorRect.width : 0;
+      break;
+    case FlyoutMenuDirection.BottomStart:
+      toReturn.top = flowDown ? 0 : -(activatorRect.height + bodyRect.height);
+      toReturn.left = flowRight ? 0 : -(bodyRect.width - activatorRect.width);
+      break;
+    default:
+  }
+
+  const offscreenRight = bodyRect.width + activatorRect.left + edgePadding > rightBound;
+  const offscreenLeft = activatorRect.right - bodyRect.width - edgePadding < 0;
+  if (flowRight && offscreenRight) {
+    const offscreenOffset = rightBound - (activatorRect.left + bodyRect.width + edgePadding);
+    toReturn.left +=
+      menuDirection === FlyoutMenuDirection.LeftEnd ||
+      menuDirection === FlyoutMenuDirection.LeftStart
+        ? -offscreenOffset
+        : offscreenOffset;
+    menuBody.children[1].style.setProperty(
+      '--after-offset',
+      `${stripUnit(rem(-offscreenOffset)) - 1}rem`
+    );
+  } else if (!flowRight && offscreenLeft) {
+    const offscreenOffset = activatorRect.right - bodyRect.width - edgePadding;
+    toReturn.left +=
+      menuDirection === FlyoutMenuDirection.LeftEnd ||
+      menuDirection === FlyoutMenuDirection.LeftStart
+        ? offscreenOffset
+        : -offscreenOffset;
+    menuBody.children[1].style.setProperty(
+      '--after-offset',
+      `${stripUnit(rem(-offscreenOffset)) - 1}rem`
+    );
+  }
+  menuBody.classList.remove(`${iotPrefix}--flyout-menu--body__${menuDirection}`);
+  menuBody.classList.add(getClassName(flowDown, flowRight));
+  return toReturn;
+};
+
+const menuOffset = (menuBody, menuDirection, menuElement, flipped) => {
+  if (!window || !document) {
+    return { top: 0, left: 0 };
+  }
+
+  if (menuBody.classList.contains(`${prefix}--tooltip`)) {
+    return tooltipOffset(menuBody, menuDirection, menuElement);
+  }
+
+  if (menuBody.classList.contains(`${prefix}--overflow-menu-options`)) {
+    return overflowMenuOffset(menuBody, menuElement, flipped);
+  }
+
+  if (menuElement.classList.contains(`${iotPrefix}--flyout-menu`)) {
+    return flyoutMenuOffset(menuBody.parentElement.parentElement, menuDirection, menuElement);
+  }
```

Comment:
This is sort of brittle - Carbon doesn't consider their classnames as part of the API that adheres to semver, so they can change them and would break this logic. Maybe it's easier to just have separate modules for flyoutPositioning, tooltipPositioning, etc?